File: 865. Smallest Subtree with all the Deepest Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        
        # find the deepest
        def helper(root):
            if not root:
                return 0
            left = helper(root.left)
            right = helper(root.right)
            return max(left, right) + 1
        deepest = helper(root)
        
        def helper2(root, current_level):
            if not root:
                return None
            if current_level == deepest:
                return root
            left = helper2(root.left, current_level+1)
            right = helper2(root.right, current_level+1)
            if left and right:
                return root
            elif left:
                return left
            elif right:
                return right
            else:
                return None
        return helper2(root, 1)
        

File: 865. Smallest Subtree with all the Deepest Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int depth;
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        depth = find_depth(root);
        
        return helper(root, 0);
    }
    
    private TreeNode helper(TreeNode node, int current_depth){
        if(node == null){
            return null;
        }
        current_depth += 1;
        if(current_depth == depth){
            return node;
        }
        TreeNode left = helper(node.left, current_depth);
        TreeNode right = helper(node.right, current_depth);
        if(left != null & right != null){
            return node;
        }
        else if(left != null){
            return left;
        }
        else if(right != null){
            return right;
        }else{
            return null;
        }
        
    }
    
    private int find_depth(TreeNode root){
        if(root == null){
            return 0;
        }
        return Math.max(find_depth(root.left), find_depth(root.right)) + 1;
    }
}